src/main/java/com/squareup/javapoet/JavaFile.java

```diff
@@ -93,7 +95,8 @@ public void writeTo(Path directory) throws IOException {
       Files.createDirectories(outputDirectory);
     }
 
-    Path outputPath = outputDirectory.resolve(typeSpec.name + ".java");
+    TypeSpec firstTypeSpec = typeSpecs.get(0);
```

Comment:
This might not be the public type

Reply:
Could it be that by convention (and documentation) file name will be the first public or package visible TypeSpec added to the JavaFile?